x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java

```diff
@@ -5,75 +5,350 @@
+            LOGGER.trace("{} no new reads, buffer limit has been reached [{}]", params.getFollowShardId(), buffer.size());
+            return false;
+        }
+        return true;
+    }
+
+    private synchronized void coordinateWrites() {
+        while (hasWriteBudget() && buffer.isEmpty() == false) {
+            if (numConcurrentWrites >= params.getMaxConcurrentWrites()) {
+                LOGGER.trace("{} maximum number of concurrent writes have been reached [{}]",
+                    params.getFollowShardId(), numConcurrentWrites);
+                break;
+            }
+            Translog.Operation[] ops = new Translog.Operation[Math.min(params.getMaxWriteSize(), buffer.size())];
+            for (int i = 0; i < ops.length; i++) {
+                ops[i] = buffer.remove();
+            }
+            numConcurrentWrites++;
+            LOGGER.trace("{}[{}] write [{}/{}] [{}]", params.getFollowShardId(), numConcurrentWrites, ops[0].seqNo(),
+                ops[ops.length - 1].seqNo(), ops.length);
+            sendBulkShardOperationsRequest(ops);
+        }
+    }
+
+    private boolean hasWriteBudget() {
+        assert Thread.holdsLock(this);
+        if (numConcurrentWrites >= params.getMaxConcurrentWrites()) {
+            LOGGER.trace("{} maximum number of concurrent writes have been reached [{}]",
+                params.getFollowShardId(), numConcurrentWrites);
+            return false;
+        }
+        return true;
+    }
+
+
+    private void sendShardChangesRequest(long from, long size, Long targetSeqNo) {
+        innerSendShardChangesRequest(from, size,
+            response -> {
+                retryCounter.set(0);
+                handleResponse(from, size, targetSeqNo, response);
+            },
+            e -> handleFailure(e, () -> sendShardChangesRequest(from, size, targetSeqNo)));
     }
 
+    private void handleResponse(long from, long size, Long targetSeqNo, ShardChangesAction.Response response) {
+        maybeUpdateMapping(response.getIndexMetadataVersion(), () -> {
+            synchronized (ShardFollowNodeTask.this) {
+                globalCheckpoint = Math.max(globalCheckpoint, response.getLeaderGlobalCheckpoint());
+                if (response.getOperations().length == 0) {
+                    if (targetSeqNo != null) {
+                        LOGGER.trace("{} received no ops while [{}/{}] was expected, re-executing read...",
+                            params.getFollowShardId(), from, size);
+                        sendShardChangesRequest(from, size, targetSeqNo);
+                    } else {
+                        numConcurrentReads--;
+                        if (numConcurrentWrites == 0) {
+                            coordinateWrites();
+                        }
+                        LOGGER.trace("{} received no ops, scheduling to coordinate reads", params.getFollowShardId());
+                        scheduler.accept(TimeValue.timeValueMillis(500), this::coordinateReads);
```

Comment:
can we make this configurable? also 500 millis is way too small and will busy spin. I guess 10s ? (the real solutions will be to have long polling, but that will come later).

Reply:
I did have to keep this delay at 1s in tests, otherwise tests take too long to complete or fail because an expected result could not be reached within assertBusy(...) calls.

Reply:
I think it can be even less in tests. No one is worried about sending multiple requests there.